File: app/src/main/java/com/popalay/cardme/business/settings/SettingsInteractor.java
package com.popalay.cardme.business.settings;

import com.popalay.cardme.data.models.Settings;
import com.popalay.cardme.data.repositories.settings.SettingsRepository;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.schedulers.Schedulers;

@Singleton
public class SettingsInteractor {

    private final SettingsRepository settingsRepository;

    @Inject public SettingsInteractor(SettingsRepository settingsRepository) {
        this.settingsRepository = settingsRepository;
    }

    public Flowable<Settings> listenSettings() {
        return settingsRepository.listen()
                .distinctUntilChanged()
                .subscribeOn(Schedulers.io());
    }

    public Flowable<Boolean> listenShowCardsBackground() {
        return listenSettings()
                .map(Settings::isCardBackground)
                .distinctUntilChanged();
    }

    public Completable saveSettings(Settings settings) {
        return settingsRepository.save(settings).subscribeOn(Schedulers.io());
    }
}

File: app/src/main/java/com/popalay/cardme/business/settings/SettingsInteractor.kt
package com.popalay.cardme.business.settings

import com.popalay.cardme.data.models.Settings
import com.popalay.cardme.data.repositories.settings.SettingsRepository
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class SettingsInteractor @Inject constructor(private val settingsRepository: SettingsRepository) {

    fun listenSettings(): Flowable<Settings> = settingsRepository.listen()
            .distinctUntilChanged()
            .subscribeOn(Schedulers.io())

    fun listenShowCardsBackground(): Flowable<Boolean> = listenSettings()
            .map { it.isCardBackground }
            .distinctUntilChanged()

    fun saveSettings(settings: Settings): Completable = settingsRepository.save(settings)
            .subscribeOn(Schedulers.io())

}
